### DISCUSSION TITLE
Upgrading to React 18 Alpha (new root API)

### DISCUSSION THREAD

**Discussion by Dmarcotrigiano** [vercel/next.js#26105] [Help]
Hello everyone, I've been looking to create a landing page for my start up using Next.js and my team has been creatively at work with Three.js. The concurrent mode on the alpha is certainly appealing and we're looking to build our prototype using that build. Unfortunately, due to the nature of Next.js it seems that the following implementation is not so simple: 

```
 useEffect(() => {
    if (typeof window !== "undefined") {
      const rootElement = document.getElementById("root");
      if (!rootElement) throw new Error("Failed to find the root element");
      const root = ReactDOM.createRoot(rootElement, { hydrate: true });
      root.render(<MyApp />);
    }
    return () => {};
  }, []);
```
      
Does anyone have an idea as to how they would migrate their Next.js application to the new createRoot API?

**Comment by NateRadebaugh** [comment]
The latest blog post for next 11 has an upgrade guide:
> https://nextjs.org/blog/next-11#upgrade-guide

> We are also working closely with the React team as they introduce React 18. Next.js 11 uses `createRoot` when React 18 alpha is being used.

However, I updated to the latest next 11 canary and the latest react alpha and I still see the following console log:
> Warning: ReactDOM.hydrate is no longer supported in React 18. Use hydrateRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot

![image](https://user-images.githubusercontent.com/130445/122280743-21681500-ceaf-11eb-9005-f158d40665a8.png)

**Reply by Dmarcotrigiano** [reply]
I noticed! I believe I started this discussion prior to the announcement.

> However, I updated to the latest next 11 canary and the latest react alpha and I still see the following console log:

The blog post simply states React 18 should be recognized and use createRoot. Is there anything that could be done to help Next.js recognize the usage of React 18, or is it that you're not using React 18 as it expects it.

**Comment by devknoll** [comment]
You need to set `experimental.reactRoot: true` in your `next.config.js` for the moment. @timneutkens I’m not sure this should be implied or implemented on the blog at the moment: it’s still alpha 😅

**Reply by Dmarcotrigiano** [reply]
Ah yes, alpha, excuse me while I push to prod.
